src/Storages/MergeTree/ReplicatedMergeTreePartCheckThread.cpp

```diff
@@ -532,6 +621,160 @@ bool ReplicatedMergeTreePartCheckThread::onPartIsLostForever(const String & part
+    constexpr auto MAX_BACKGROUND_CHECK_BACKOFF_TIMEOUT_MS = std::chrono::milliseconds{10 * 60 * 1000}; // 10 minutes
+
+    auto backoff_timeout_ms = background_check_backoff_timeout_ms;
+    if (backoff_timeout_ms.count())
+    {
+        backoff_timeout_ms *= 2;
+        if (backoff_timeout_ms > MAX_BACKGROUND_CHECK_BACKOFF_TIMEOUT_MS)
+            backoff_timeout_ms = MAX_BACKGROUND_CHECK_BACKOFF_TIMEOUT_MS;
+    }
+    else
+        backoff_timeout_ms = MIN_BACKGROUND_CHECK_BACKOFF_TIMEOUT_MS;
+
+    background_check_backoff_timeout_ms = backoff_timeout_ms;
+    return background_check_backoff_timeout_ms;
+}
+
+std::chrono::milliseconds ReplicatedMergeTreePartCheckThread::delayBetweenChecks()
+{
+    /// backoff timeout is used to avoid overscheduling in case there is not parts to check and background_part_check_delay_seconds is small
+    using namespace std::chrono;
+    const auto background_part_check_delay_ms
+        = duration_cast<milliseconds>(seconds{storage.getSettings()->background_part_check_delay_seconds});
+    const auto backoff_timeout_ms = updateBackgroundCheckBackoffTimeout();
+    return std::max(background_part_check_delay_ms, backoff_timeout_ms);
+}
+
+void ReplicatedMergeTreePartCheckThread::doBackgroundPartCheck()
+{
+    using namespace std::chrono;
+    auto part_to_check = choosePartForBackgroundCheck();
+    if (!part_to_check)
+    {
+        const auto delay_between_checks = delayBetweenChecks();
+        enqueueBackgroundPartCheck(delay_between_checks);
+        return;
+    }
+
+    /// reset backoff timeout, there is a part to check
+    background_check_backoff_timeout_ms = milliseconds{0};
+
+    try
+    {
+        auto table_lock = storage.lockForShare(RWLockImpl::NO_QUERY, storage.getSettings()->lock_acquire_timeout_for_background_operations);
+
+        LOG_TRACE(log, "Background part check: going to check part {}", part_to_check->name);
+
+        last_checked_part = part_to_check->info;
+        auto check_start = steady_clock::now();
+        auto result = checkActivePart(part_to_check);
+        auto check_duration_ms = duration_cast<milliseconds>(steady_clock::now() - check_start);
+
+        part_to_check->last_check_time = steady_clock::now();
+
+        LOG_TRACE(log, "Background part check took {} ms", check_duration_ms.count());
+
+        if (result.status.success)
+        {
+            LOG_INFO(
+                log,
+                "Background part check succeeded: part={} path={} took={}ms",
+                part_to_check->name,
+                result.status.fs_path,
+                check_duration_ms.count());
+        }
+        else
+        {
+            LOG_WARNING(
+                log,
+                "Background part check failed: part={} path={} message={}",
+                part_to_check->name,
+                result.status.fs_path,
+                result.status.failure_message);
+        }
```

Comment:
What's the point of the background parts check if we only print a warning in case of failure?
However, it's the safest option, and it still may be useful to set up a log-based alert...